File: tests/algorithms/quicksort/quicksort.py
#!/usr/bin/env python3
"""
Quicksort benchmark implementation in Python.
Sorts an array using the quicksort algorithm.
"""

import sys
import time
import random


def quicksort(arr):
    """Sort array using quicksort algorithm."""
    if len(arr) <= 1:
        return arr
    
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    
    return quicksort(left) + middle + quicksort(right)


def main():
    """Main execution function."""
    # Generate test data
    size = 10000
    arr = list(range(size))
    random.shuffle(arr)
    
    print(f"Sorting array of size {size}...")
    start_time = time.perf_counter()
    
    sorted_arr = quicksort(arr.copy())
    
    end_time = time.perf_counter()
    execution_time = end_time - start_time
    
    # Verify correctness
    is_sorted = sorted_arr == sorted(arr)
    
    print(f"Result: {'Sorted correctly' if is_sorted else 'Sort failed'}")
    print(f"Execution time: {execution_time:.6f} seconds")


if __name__ == "__main__":
    main()

File: tests/algorithms/quicksort/quicksort.ts
#!/usr/bin/env node
/**
 * Quicksort benchmark implementation in TypeScript.
 * Sorts an array using the quicksort algorithm.
 */

function quicksort(arr: number[]): number[] {
    if (arr.length <= 1) {
        return arr;
    }
    
    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);
    
    return [...quicksort(left), ...middle, ...quicksort(right)];
}

function shuffle(arr: number[]): void {
    for (let i = arr.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [arr[i], arr[j]] = [arr[j], arr[i]];
    }
}

function main(): void {
    const size = 10000;
    const arr = Array.from({length: size}, (_, i) => i);
    shuffle(arr);
    
    console.log(`Sorting array of size ${size}...`);
    const start = performance.now();
    
    const sortedArr = quicksort([...arr]);
    
    const end = performance.now();
    const executionTime = (end - start) / 1000; // Convert to seconds
    
    // Verify correctness
    const expected = [...arr].sort((a, b) => a - b);
    const isSorted = JSON.stringify(sortedArr) === JSON.stringify(expected);
    
    console.log(`Result: ${isSorted ? 'Sorted correctly' : 'Sort failed'}`);
    console.log(`Execution time: ${executionTime.toFixed(6)} seconds`);
}

// Always run main for benchmark purposes
main();